pkg/cloud/aws/actuators/machine/utils.go

```diff
@@ -141,30 +141,40 @@ func terminateInstances(client awsclient.Client, instances []*ec2.Instance) erro
 func ProviderConfigFromMachine(client client.Client, machine *clusterv1.Machine, codec *providerconfigv1.AWSProviderConfigCodec) (*providerconfigv1.AWSMachineProviderConfig, error) {
 	var providerConfig runtime.RawExtension
 
-	if machine.Spec.ProviderConfig.Value == nil && machine.Spec.ProviderConfig.ValueFrom == nil {
+	// TODO(jchaloup): Remove providerConfig once all consumers migrate to providerSpec
+	providerSpec := machine.Spec.ProviderConfig
+	// providerSpec has higher priority over providerConfig
+	if machine.Spec.ProviderSpec.Value != nil || machine.Spec.ProviderSpec.ValueFrom != nil {
+		providerSpec = machine.Spec.ProviderSpec
+		fmt.Printf("Falling to default providerSpec\n")
+	} else {
+		fmt.Printf("Falling to providerConfig\n")
+	}
+
+	if providerSpec.Value == nil && providerSpec.ValueFrom == nil {
 		return nil, fmt.Errorf("unable to find machine provider config: neither Spec.ProviderConfig.Value nor Spec.ProviderConfig.ValueFrom set")
 	}
 
-	// If no machine.Spec.ProviderConfig.Value then we lookup for machineClass
-	if machine.Spec.ProviderConfig.Value != nil {
-		providerConfig = *machine.Spec.ProviderConfig.Value
+	// If no providerSpec.Value then we lookup for machineClass
+	if providerSpec.Value != nil {
+		providerConfig = *providerSpec.Value
 	} else {
-		if machine.Spec.ProviderConfig.ValueFrom.MachineClass == nil {
+		if providerSpec.ValueFrom.MachineClass == nil {
 			return nil, fmt.Errorf("unable to find MachineClass on Spec.ProviderConfig.ValueFrom")
 		}
 		machineClass := &clusterv1.MachineClass{}
 		key := types.NamespacedName{
-			Namespace: machine.Spec.ProviderConfig.ValueFrom.MachineClass.Namespace,
-			Name:      machine.Spec.ProviderConfig.ValueFrom.MachineClass.Name,
+			Namespace: providerSpec.ValueFrom.MachineClass.Namespace,
+			Name:      providerSpec.ValueFrom.MachineClass.Name,
 		}
 		if err := client.Get(context.Background(), key, machineClass); err != nil {
 			return nil, err
 		}
-		providerConfig = machineClass.ProviderConfig
+		providerConfig = machineClass.ProviderSpec
 	}
 
 	var config providerconfigv1.AWSMachineProviderConfig
-	if err := codec.DecodeProviderConfig(&clusterv1.ProviderConfig{Value: &providerConfig}, &config); err != nil {
+	if err := codec.DecodeProviderConfig(&clusterv1.ProviderSpec{Value: &providerConfig}, &config); err != nil {
```

Comment:
Change `{Value: &providerConfig...` to `providerSpec`?

Reply:
Bad naming of variables I guess. `providerConfig` variable is defined as  `var providerConfig runtime.RawExtension`. So it's actually of type `runtime.RawExtension`.

Reply:
variable renamed to `providerSpecRawExtention`